xdsl/utils/parse_pipeline.py

```diff
@@ -0,0 +1,227 @@
+            for pattern, kind in _lexer_rules:
+                if (match := pattern.match(input_str, pos)) is not None:
+                    token = Token(Span(match.start(), match.end(), input), kind)
+                    pos = match.end()
+                    break
+            if token is None:
+                raise PassPipelineParseError(
+                    Token(Span(pos, pos + 1, input), Kind.IDENT), "Unknown token"
+                )
+            yield token
+            if pos >= end:
+                yield Token(Span(pos, pos + 1, input), Kind.EOF)
+                return
+
+    def lex(self) -> Token:
+        token = self.peek()
+        self._peeked = None
+        return token
+
+    def peek(self) -> Token:
+        if self._peeked is None:
+            self._peeked = next(self._stream)
+        return self._peeked
+
+
+class PassPipelineParseError(BaseException):
+    def __init__(self, token: Token, msg: str):
+        super().__init__(
+            "Error parsing pass pipeline specification:\n"
+            + token.span.print_with_context(msg)
+        )
+
+
+_PassArgTypes = list[str | int | bool | float]
+
+
+def parse_pipeline(
+    pipeline_spec: str,
+) -> Iterator[tuple[str, dict[str, _PassArgTypes]]]:
+    """
+    This takes a pipeline string and gives a representation of
+    the specification.
+
+    Each pass is represented by a tuple of:
+     - name: the name of the pass as string
+     - args: a dictionary, where each value is zero or more
+            of (str | bool | float | int)
+    """
+    lexer = PipelineLexer(pipeline_spec)
+
+    while True:
+        name = lexer.lex()
+        if name.kind is Kind.EOF:
+            return
+        if name.kind is not Kind.IDENT:
+            raise PassPipelineParseError(name, "Expected pass name here")
+
+        delim = lexer.lex()
+        if delim.kind is Kind.EOF:
+            yield name.span.text, dict()
+            return
+
+        if delim.kind is Kind.COMMA:
+            yield name.span.text, dict()
+            continue
+
+        if delim.kind is not Kind.L_BRACE:
+            raise PassPipelineParseError(
+                delim, "Expected a comma or pass arguments here"
+            )
+
+        yield name.span.text, _parse_pass_args(lexer)
+
+        delim = lexer.lex()
+        if delim.kind is Kind.EOF:
+            return
+        if delim.kind is not Kind.COMMA:
+            raise PassPipelineParseError(
+                delim, "Expected a comma after pass argument dict here"
+            )
+
+
+def _parse_pass_args(lexer: PipelineLexer):
+    args: dict[str, _PassArgTypes] = dict()
+
+    while True:
+        name = lexer.lex()
+
+        # allow for zero-length arg dicts
+        if name.kind is Kind.R_BRACE:
+            return args
+
+        if name.kind is not Kind.IDENT:
+            raise PassPipelineParseError(name, "Expected argument name here")
+
+        # handle zero-length args
+        if lexer.peek().kind in (Kind.SPACE, Kind.R_BRACE):
```

Comment:
If you get a space here, you don't consume it, and I think you should?

Reply:
Maybe just `if `lexer.peek().kind` is not Kind.EQUALS:`?
I feel that you are often encoding the information that you know the next tokens, while actually you don't need to. If you just encode what you expect in your current rule, that is usually enough.